homeassistant/components/mobile_app/notify.py

```diff
@@ -0,0 +1,117 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = ["mobile_app"]
+
+
+# pylint: disable=invalid-name
+def log_rate_limits(hass, device_name, resp, level=20):
+    """Output rate limit log line at given level."""
+    rate_limits = resp["rateLimits"]
+    resetsAt = dt_util.parse_datetime(rate_limits["resetsAt"])
+    resetsAtTime = resetsAt - datetime.now(timezone.utc)
+    rate_limit_msg = ("mobile_app push notification rate limits for %s: "
+                      "%d sent, %d allowed, %d errors, "
+                      "resets in %s")
+    _LOGGER.log(level, rate_limit_msg,
+                device_name,
+                rate_limits["successful"],
+                rate_limits["maximum"], rate_limits["errors"],
+                str(resetsAtTime).split(".")[0])
+
+
+def get_service(hass, config, discovery_info=None):
+    """Get the mobile_app notification service."""
+    return MobileAppNotificationService()
+
+
+class MobileAppNotificationService(BaseNotificationService):
+    """Implement the notification service for mobile_app."""
+
+    def __init__(self):
+        """Initialize the service."""
+
+    @property
+    def targets(self):
+        """Return a dictionary of registered targets."""
+        return push_registrations(self.hass)
+
+    def send_message(self, message="", **kwargs):
+        """Send a message to the Lambda APNS gateway."""
+        data = {ATTR_MESSAGE: message}
+
+        if kwargs.get(ATTR_TITLE) is not None:
+            # Remove default title from notifications.
+            if kwargs.get(ATTR_TITLE) != ATTR_TITLE_DEFAULT:
+                data[ATTR_TITLE] = kwargs.get(ATTR_TITLE)
+
+        targets = kwargs.get(ATTR_TARGET)
+
+        if not targets:
+            targets = push_registrations(self.hass)
+
+        if kwargs.get(ATTR_DATA) is not None:
+            data[ATTR_DATA] = kwargs.get(ATTR_DATA)
+
+        for target in targets:
+
+            entry = self.hass.data[DOMAIN][DATA_CONFIG_ENTRIES][target]
+            entry_data = entry.data
+
+            app_data = entry_data[ATTR_APP_DATA]
+            push_token = app_data[ATTR_PUSH_TOKEN]
+            push_url = app_data[ATTR_PUSH_URL]
+
+            data[ATTR_PUSH_TOKEN] = push_token
+
+            reg_info = {
+                ATTR_APP_DATA: app_data,
+                ATTR_APP_ID: entry_data[ATTR_APP_ID],
+                ATTR_APP_NAME: entry_data[ATTR_APP_NAME],
+                ATTR_APP_VERSION: entry_data[ATTR_APP_VERSION],
+                ATTR_DEVICE_NAME: entry_data[ATTR_DEVICE_NAME],
+                ATTR_MANUFACTURER: entry_data[ATTR_MANUFACTURER],
+                ATTR_MODEL: entry_data[ATTR_MODEL],
+                ATTR_OS_NAME: entry_data[ATTR_OS_NAME],
+                CONF_WEBHOOK_ID: entry_data[CONF_WEBHOOK_ID],
+            }
+            if ATTR_OS_VERSION in entry_data:
+                reg_info[ATTR_OS_VERSION] = entry_data[ATTR_OS_VERSION]
+
+            data['registration_info'] = reg_info
```

Comment:
Why do we include registration info in the push ? Why would the app name ever need to know it's own ID, webhook ID etc ?

Reply:
This goes to the server which may not have all the details and may want them.

Reply:
Do you have a use case? Generally we aim to keep the server side just a passthrough.

Also, do we encrypt the data ?

Reply:
I do have a use case. Firebase doesn't allow us to set any device specific data alongside the push registration. To enable using the old iOS-Push/SNS logic to properly build the payload, I need to inspect the app ID to see if its the iOS bundle.

Reply:
By data, do you mean the registration data or the entire notification payload sent to the server?

Reply:
both.

Reply:
Coming back to the encryption part of this specifically: I absolutely want to implement it, but it will be a bit tough to do in a generic way as some fields can be encrypted while others, such as those that the device automatically interprets or the server needs to be able to do its work, cannot be encrypted. So we'd need to allow apps to provide a schema of fields that can be encrypted I guess? Alternatively, do we just always encrypt the `title` and `message` and anything in `data` is just left unencrypted. That's fine for a MVP, but at least in the case of iOS, `data` probably has the most sensitive contents such as image URLs, entity IDs and latitude and longitude.